### ISSUE TITLE
Logging adjustment to support more schemas

### ISSUE THREAD

**Issue by dlangerm** [PyTorchLightning/lightning-flash#818]
## 🚀 Feature
Lightning supports weights & biases but the default logging behavior isn't very conducive to the way the charts are organized. Additionally, it doesn't properly support MetricCollections from torchmetrics.

Specifically, lightning's log schema is: 

`<stage>_<metric>_<epoch/step>`

W&B supports separating charts based on `/` delimited strings. So in my modules I've resorted to overriding the default training_step with the following snippet. 

```
def training_step(self, batch: Any, batch_idx: int) -> Any:
      batch = (batch[DefaultDataKeys.INPUT], batch[DefaultDataKeys.TARGET])
      output = self.step(batch, batch_idx, self.train_metrics)
      
      # support metric collections, would need a check here for a general fix
      self.log_dict(
          {f"{k}/train": v for k,
              v in output["logs"]['metriccollection'].items()},
          on_step=True, on_epoch=True, prog_bar=True
      )
      
      # support non-metric collections, with w&b schema
      self.log_dict(
          {f"{k}/train": v for k,
              v in output["logs"].items() if k != 'metriccollection'},
          on_step=True, on_epoch=True, prog_bar=True
      )
      return output["loss"]
```

### Motivation

I like to use weights & biases and having better integrated support would be pretty cool.

### Pitch

Obviously, the above snippet isn't exactly pretty, but it is functional. A better way would probably be to abstract this into a call on a Task 

```
def __init__(..., log_schema=default_log_schema):
   ...
   if isinstance(log_schema, str): 
            # use schema coded in flash
   else:
          # expect some kind of callable for train/val/test
          log_schema = 'custom' 
          
   self.log_schema = log_schema # probably want some kind of enum to avoid hardcoding these strings
   self._log_schemas = {
        'default': {
         populate with default flash code
        },
        'wandb': {
        above functions, tuned for train/val/test
        },
        'custom': {
        dict of callables for [str] -> dict where dict is what we actually pass to log_dict
        }
   }

def get_log_dict(output, stage):
     return self._log_schemas[self.log_schema](output, stage)
     
def training_step(self, batch: Any, batch_idx: int) -> Any:
    batch = (batch[DefaultDataKeys.INPUT], batch[DefaultDataKeys.TARGET])
    output = self.step(batch, batch_idx, self.train_metrics)
    self.log_dict(self.get_log_dict(output), 'train') # or some kind of enum for train/val/test
    return output["loss"]
```

### Alternatives

Overriding the training step is fine, except if an intermediate class existed, needing to override `x_step` gets tedious. Another alternative would be to pass in hooks to manipulate the logs before they're passed to log_dict, but that would necessitate more of an interface change and could potentially break down the line if log_dict ever changes.

### Additional context

This also reduces the hardcoding of the log format and allows anyone to produce their own log formats.
